File: go/linkedlist.go
package linkedlist

import "fmt"

type Node[T any] struct {
	Value    T
	Next     *Node[T]
	Previous *Node[T]
}

type LinkedList[T any] struct {
	Head   *Node[T]
	Equals func(a, b T) bool
	Size   int
}

func (l *LinkedList[T]) Append(value T) {
	newNode := &Node[T]{Value: value}
	if l.Head == nil {
		l.Head = newNode
		l.Size++
		return
	}

	current := l.Head
	for current.Next != nil {
		current = current.Next
	}
	current.Next = newNode
	l.Size++
}

func (l *LinkedList[T]) Prepend(value T) {
	newNode := &Node[T]{Value: value, Next: l.Head}
	l.Head = newNode
	l.Size++
}

func (l *LinkedList[T]) Traverse() {
	current := l.Head
	for current != nil {
		fmt.Println(current.Value)
		current = current.Next
	}
	fmt.Println("")
}

func (l *LinkedList[T]) Search(value T) *Node[T] {
	if l.Head == nil || l.Equals == nil {
		return nil
	}

	current := l.Head
	for current != nil {
		if l.Equals(current.Value, value) {
			return current
		}
		current = current.Next
	}
	return nil
}

func (l *LinkedList[T]) Pop() (T, bool) {
	if l.Head == nil {
		var zeroValue T
		return zeroValue, false
	}

	if l.Head.Next == nil {
		value := l.Head.Value
		l.Head = nil
		l.Size--
		return value, true
	}

	current := l.Head
	for current.Next.Next != nil {
		current = current.Next
	}

	value := current.Next.Value
	current.Next = nil
	l.Size--
	return value, true
}

func (l *LinkedList[T]) Dequeue() (T, bool) {
	if l.Head == nil {
		var zeroValue T
		return zeroValue, false
	}

	value := l.Head.Value
	l.Head = l.Head.Next
	l.Size--
	return value, true
}

File: python/linked-list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, value):
        new = Node(value)
        if not self.head:
            self.head = new
            return
        last = self.head
        while last.next:
            last = last.next
        last.next = new

    def remove(self, value):
        current = self.head
        if current and current.value == value:
            self.head = current.next
            current = None
            return
        previous = None
        while current and current.value != value:
            previous = current
            current = current.next
        if not current:
            print("value not found")
            return
        previous.next = current.next
        current = None

    def print(self):
        current = self.head
        while current:
            print(current.value, end=" -> ")
            current = current.next
        print("None")

if __name__ == "__main__":
    list = LinkedList()
    for num in [24, 83, 11, 56, 2, 71, 39, 48, 65, 90]:
        list.append(num)
    list.print()
    for num in [83, 56, 39]:
        list.remove(num)
    list.print()
